Core/src/org/sleuthkit/autopsy/datamodel/utils/BackgroundTaskRunner.java

```diff
@@ -0,0 +1,163 @@
+     * supplied by this utility.
+     */
+    @FunctionalInterface
+    public interface NodeTask {
+
+        /**
+         * Performs a task in a thread supplied by this utility.
+         *
+         * @param future The future of the actual background task executing this
+         *               method, may be checked for task cancellation.
+         *
+         * @return A PropertyChangeEvent holding the result of the task.
+         *
+         * @throws Exception If there is an error performing the task.
+         */
+        PropertyChangeEvent run(Future<?> future) throws Exception;
+    }
+
+    /**
+     * Submits a background task for an AbstractNode to a dedicated thread pool.
+     *
+     * @param node     The AbstractNode.
+     * @param task     The task to be done in the background.
+     * @param listener A PropertyChangeListener for the AbstractNode to handle
+     *                 the PropertyChangeEvent produced by the task.
+     *
+     * @return The Future for the Runnable used to run the task.
+     */
+    public static Future<?> submitTask(AbstractNode node, NodeTask task, PropertyChangeListener listener) {
+        NodeBackgroundTask backgroundTask = new NodeBackgroundTask(node, task, listener);
+        Future<?> future = executor.submit(backgroundTask);
+        backgroundTask.setFuture(future);
+        return future;
+    }
+
+    /**
+     * A Runnable that uses weak references to an AbstractNode and the
+     * PropertyChangeListener for the node to safely return results to the node,
+     * if it has not been destroyed by the NetBeans framework at the time the
+     * task is completed.
+     */
+    private static class NodeBackgroundTask implements Runnable {
+
+        private final WeakReference<AbstractNode> weakNodeRef;
+        private final NodeTask task;
+        private final PropertyChangeListener weakListenerRef;
+        private Future<?> future;
+
+        /**
+         * Constructs a Runnable that uses weak references to an AbstractNode
+         * and the PropertyChangeListener for the node to safely return results
+         * to the node, if it still exists at the time the task is completed.
+         *
+         * @param node     The AbstractNode.
+         * @param task     The task to be done in the background.
+         * @param listener A PropertyChangeListener for the AbstractNode to
+         *                 handle the PropertyChangeEvent produced by the task.
+         */
+        private NodeBackgroundTask(AbstractNode node, NodeTask task, PropertyChangeListener listener) {
+            this.weakNodeRef = new WeakReference<>(node);
+            this.task = task;
+            this.weakListenerRef = WeakListeners.propertyChange(listener, null);
+        }
+
+        @Override
+        public void run() {
+            AbstractNode node = weakNodeRef.get();
+            if (node == null) {
+                return;
+            }
+
+            if (future.isCancelled()) {
```

Comment:
For instance, this could be a check to Thread.currentThread().isInterrupted(). Then the problem of passing the Future before execution is gone.

Reply:
I have used that technique before. I agree that it is a good simplification. It just means that the task client doing the cancellation needs to call Future.cancel(true).